Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,358 @@
+
+* For linear maps `f : M →ₗ[R] N`  and `g : N →ₗ[R] P`, `Exact f g` says that `range f = ker g``
+
+
+* `TensorProduct.rTensor_exact` says that one can tensor a short exact sequence on the right
+
+
+* `TensorProduct.lTensor_exact` says that one can tensor a short exact sequence on the left
+
+* For `N : Submodule R M`, `LinearMap.exact_subtype_mkQ N` says that
+  the inclusion of the submodule and the quotient map form an exact pair,
+  and `lTensor_mkQ` compute `ker (lTensor Q (N.mkQ))` and similarly for `rTensor_mkQ`
+
+* `TensorProduct.map_ker` computes the kernel of `TensorProduct.map g g'`
+in the presence of two short exact sequences.
+
+The proofs are those of [bourbaki1989] (chap. 2, §3, n°6)
+
+
+## TODO
+
+
+* Analogue for morphisms of algebras
+
+* Treat the noncommutative case
+
+* Treat the case of modules over semirings
+(For a possible definition of an exact sequence of commutative semigroups, see
+  [Grillet-1969b], Pierre-Antoine Grillet,
+  *The tensor product of commutative semigroups*,
+  Trans. Amer. Math. Soc. 138 (1969), 281-293, doi:10.1090/S0002-9947-1969-0237688-1 .)
+
+-/
+
+section Exact
+
+variable {M N P : Type _} (f : M → N) (g : N → P)
+
+def Exact [Zero P] : Prop := ∀ y, g y = 0 ↔ y ∈ Set.range f
+
+lemma Exact.comp_eq_zero [Zero P] (h : Exact f g) : g.comp f = 0 := by
+  ext x
+  simp only [Function.comp_apply, Pi.zero_apply]
+  rw [h]
+  exact Set.mem_range_self x
+
+open LinearMap
+
+variable {R : Type _} [CommSemiring R]
+variable {M N P : Type _}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommMonoid P]
+  [Module R M] [Module R N] [Module R P]
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+lemma Exact.linearMap_ker_eq (hfg : Exact f g) : ker g = range f := by
+  ext y
+  exact hfg y
+
+lemma LinearMap.exact_iff : Exact f g ↔ LinearMap.ker g = LinearMap.range f := by
+  constructor
+  · intro h; apply Exact.linearMap_ker_eq; exact h
+  · intro h y; rw [← mem_ker, h]; rfl
+
+lemma Exact.linearMap_comp_eq_zero (h : Exact f g) : g.comp f = 0 := by
+  ext x
+  simp only [coe_comp, Function.comp_apply, zero_apply]
+  rw [h]
+  exact Set.mem_range_self x
+
+
+end Exact
+
+section Modules
+
+open TensorProduct LinearMap
+
+section Semiring
+
+variable {R : Type _} [CommSemiring R]
+variable {M N P P' : Type _}
+  [AddCommMonoid M] [AddCommMonoid N] [AddCommGroup P] [AddCommMonoid P']
+  [Module R M] [Module R N] [Module R P] [Module R P']
+
+variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
+
+variable {Q : Type _} [AddCommMonoid Q] [Module R Q]
+
+variable (g)
+
+lemma le_comap_range_lTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (lTensor Q g)).comap (TensorProduct.mk R Q P q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨q ⊗ₜ[R] n, rfl⟩
+
+lemma le_comap_range_rTensor (q : Q) :
+    LinearMap.range g ≤ (LinearMap.range (rTensor Q g)).comap
+      ((TensorProduct.mk R P Q).flip q) := by
+  rintro x ⟨n, rfl⟩
+  exact ⟨n ⊗ₜ[R] q, rfl⟩
+
+variable (Q) {g}
+
+theorem rTensor.Surjective (hg : Function.Surjective g) :
```

Comment:
I think the naming guide would suggest
```suggestion
theorem rTensor_surjective (hg : Function.Surjective g) :
```

Reply:
I make the change, but what about `Function.Surjective` (with a capital S)?

Reply:
`Function.Surjective` is a type (Prop) while `rTensor_surjective` is a term (proof) of that type.